File: ios-native/Leyne/V2/Proximity.swift
// Proximity & occupancy.
//
// ETA / soon-ness is deliberately NOT colour-coded: arrival times read as
// uniform ink, and only confidence dims them (scheduled / ghost arrivals show
// faint — the honesty whisper, see Confidence.swift). The one remaining colour
// signal here is occupancy — how full a bus is (seats green · standing amber ·
// limited grey).
//
// `ETATier` survives only to drive the "Arriving soon" text cue for an
// imminent *live* arrival; it no longer maps to a colour.

import SwiftUI

// MARK: - ETA proximity

/// How soon an arrival is, bucketed for the "Arriving soon" copy on the lead
/// chip. Thresholds mirror the mockups: ≤~2.5 min reads as imminent.
enum ETATier {
    case imminent   // arriving / very soon → "Arriving soon"
    case soon
    case medium
    case far

    static func of(etaSec: Int) -> ETATier {
        switch etaSec {
        case ..<150:  return .imminent
        case ..<540:  return .soon
        case ..<960:  return .medium
        default:      return .far
        }
    }

    var isImminent: Bool { self == .imminent }
}

// MARK: - Clock

/// Wall-clock time string honouring the in-app 24-hour preference
/// (`AppModel.use24h`). Mirrors HomeView.formattedTime so route times read the
/// same as the rest of the app. Used for per-stop ETAs on the Bus view.
func fmtClock(_ date: Date, use24h: Bool) -> String {
    let cal = Calendar.current
    let hour = cal.component(.hour, from: date)
    let minute = cal.component(.minute, from: date)
    let mm = String(format: "%02d", minute)
    if use24h {
        return "\(String(format: "%02d", hour)):\(mm)"
    }
    let h12 = ((hour + 11) % 12) + 1
    return "\(h12):\(mm) \(hour < 12 ? "AM" : "PM")"
}

// MARK: - Occupancy

/// Shared crowding colour: seats → green, standing → amber, limited/unknown
/// → neutral grey. Used by both `OccupancyLabel` (icon + words) and
/// `CrowdMeter` (bars) so the two read the same.
func occupancyColor(_ load: Load?, t: Theme) -> Color {
    switch load {
    case .sea:        return t.soon
    case .sda:        return t.mid
    case .lsd, .none: return t.dim
    }
}

// MARK: - Occupancy label

/// Crowding shown the friendly way — an icon + plain words, coloured by how
/// much room is left: seats (green) · standing (amber) · limited (grey).
/// Sits under the destination on Stop/Bus rows. (The compact `CrowdMeter`
/// bars stay for the Bus view's tighter header.)
struct OccupancyLabel: View {
    let load: Load
    let t: Theme
    var size: CGFloat = 12

    private var icon: String {
        switch load {
        case .sea: return "chair.lounge"
        case .sda, .lsd: return "person.2.fill"
        }
    }

    private var text: String {
        switch load {
        case .sea: return "Seats available"
        case .sda: return "Standing available"
        case .lsd: return "Limited standing"
        }
    }

    private var tint: Color { occupancyColor(load, t: t) }

    var body: some View {
        HStack(spacing: 5) {
            Image(systemName: icon)
                .font(.system(size: size, weight: .semibold))
            Text(text)
                .font(t.sans(size, weight: .medium))
        }
        .foregroundStyle(tint)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(text)
    }
}

File: lib/widgets/v2/proximity.dart
// Proximity & occupancy.
// Dart port of ios-native/Leyne/V2/Proximity.swift.
//
// ETA / soon-ness is deliberately NOT colour-coded: arrival times read as
// uniform ink, and only confidence dims them (scheduled/ghost arrivals show
// faint — the honesty whisper, see confidence.dart). The one remaining colour
// signal here is occupancy — how full a bus is (seats green · standing amber ·
// limited grey).
//
// (iOS keeps an `ETATier` enum to drive an "Arriving soon" text cue on its
// Home mini-chips; the Flutter UI has no such cue, so it isn't ported here.)

import 'package:flutter/material.dart';

import '../../data/models.dart';
import '../../theme.dart';
import 'confidence.dart';

// ─── ETA proximity ─────────────────────────────────────────────────────────

/// Ink for an arrival's ETA. ETA numerals always render at full foreground ink
/// regardless of confidence — the app sells timeliness, not uncertainty.
/// The one exception preserved: an imminent live arrival uses t.accent (handled
/// at each call site via `arriving && conf == ArrivalConfidence.live`).
Color etaColor({
  required int etaSec,
  required ArrivalConfidence confidence,
  required LyneTheme t,
}) => t.fg;

// ─── Occupancy ─────────────────────────────────────────────────────────────

/// Shared crowding colour: seats → green, standing → amber, limited/unknown
/// → neutral grey. Used by OccupancyLabel (CrowdMeter has its own copy to
/// avoid a circular import).
///
/// Colours are HARDCODED (not sourced from t.soon/t.mid) so occupancy stays
/// green/amber after the 2.6.0 monochrome theme change. Mirrors the intent
/// of iOS Proximity.swift occupancyColor, where t.soon/t.mid are also the
/// source but iOS keeps them coloured — Android makes this explicit instead.
///   sea → system green   (seats available)
///   sda → system orange  (standing available)
///   lsd/null → t.dim     (limited/unknown — neutral grey)
Color occupancyColor(Load? load, LyneTheme t) {
  switch (load) {
    case Load.sea:
      return const Color(0xFF34C759); // system green
    case Load.sda:
      return const Color(0xFFFF9500); // system orange/amber
    case Load.lsd:
    case null:
      return t.dim;
  }
}

// ─── OccupancyLabel ────────────────────────────────────────────────────────

/// Crowding label — icon + plain-English words, coloured by space remaining.
/// seats (green) · standing (amber) · limited (grey).
/// Sits under the destination on Stop/Bus rows.
class OccupancyLabel extends StatelessWidget {
  const OccupancyLabel({super.key, required this.load, this.size = 12});

  final Load load;
  final double size;

  @override
  Widget build(BuildContext context) {
    final t = context.t;
    final color = occupancyColor(load, t);
    return Semantics(
      label: _text,
      excludeSemantics: true,
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Icon(_icon, size: size, color: color),
          const SizedBox(width: 4),
          Text(
            _text,
            style: t.mono(size, weight: FontWeight.w500, color: color),
          ),
        ],
      ),
    );
  }

  IconData get _icon {
    switch (load) {
      case Load.sea:
        return Icons.weekend_outlined; // seats
      case Load.sda:
        return Icons.people_outline;
      case Load.lsd:
        return Icons.people;
    }
  }

  String get _text {
    switch (load) {
      case Load.sea:
        return 'Seats available';
      case Load.sda:
        return 'Standing available';
      case Load.lsd:
        return 'Limited standing';
    }
  }
}
